packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1970 @@
+   * error as the reason. This method is effectively a wrapper around a
+   * messenger call for the `ApprovalController:rejectRequest` action.
+   *
+   * @see {@link PermissionController.acceptPermissionsRequest} and
+   * {@link PermissionController.rejectPermissionsRequest} for usage.
+   * @param id - The id of the request to reject.
+   * @param error - The error associated with the rejection.
+   */
+  private _rejectPermissionsRequest(id: string, error: Error): void {
+    return this.messagingSystem.call(
+      'ApprovalController:rejectRequest',
+      id,
+      error,
+    );
+  }
+
+  /**
+   * Executes a restricted method as the subject with the the given origin.
+   * The specified params, if any, will be passed to the method implementation.
+   *
+   * **ATTN:** Great caution should be exercised in the use of this method.
+   * Methods that cause side effects or affect application state should
+   * be avoided.
+   *
+   * This method will first attempt to retrieve the requested restrictded method
+   * implementation, throwing if it does not exist. The method will then be
+   * invoked as though the subject with the specified origin had invoked it with
+   * the specified parameters. This means that any existing caveats will be
+   * applied to the restricted method, and this method will throw if the
+   * restricted method or its caveat decorators throw.
+   *
+   * In addition, this method will throw if the subject does not have a
+   * permission for the specified restricted method.
+   *
+   * @param origin - The origin of the subject to execute the method on behalf
+   * of.
+   * @param methodName - The name of the method to execute.
+   * @param params - The parameters to pass to the method implementation.
+   * @returns The result of the executed method.
+   */
+  async executeRestrictedMethod(
+    origin: OriginString,
+    methodName: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    params?: RestrictedMethodParameters,
+  ): Promise<Json> {
+    // Throws if the method does not exist
+    const methodImplementation = this._getRestrictedMethod(methodName, origin);
+
+    const result = await this._executeRestrictedMethod(
+      methodImplementation,
+      { origin },
+      methodName,
+      params,
+    );
+
+    if (result === undefined) {
+      throw new Error(
+        `Internal request for method "${methodName}" as origin "${origin}" returned no result.`,
+      );
+    }
+
+    return result;
+  }
+
+  /**
+   * An internal method used in the controller's `json-rpc-engine` middleware
+   * and {@link PermissionController.executeRestrictedMethod}. Calls the
+   * specified restricted method implementation after decorating it with the
+   * caveats of its permission. Throws if the subject does not have the
+   * requisite permission.
+   *
+   * **ATTN:** Assumes that all parameters are valid.
```

Comment:
Nit: At first I took this to mean validation happened before this call. This might be clearer if we say who is responsible, rather than saying it's not done here.

```suggestion
   * **ATTN:** Parameter validation is the responsibility of the method implementations.
```

Reply:
Hrm, well, the responsibility of the middleware and the method implementations. I'll take a stab at it locally.

Reply:
47ff014